### PR TITLE
Feat(eos_designs): Underlay OSPF authentication

### PR DESCRIPTION
**Pull request by jonxstill**
## Change Summary

Automatic encryption of underlay OSPF message digest keys for use in underlay.

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes

Because OSPF message-digest keys are encrypted in EOS configurations, using the interface name as the key, it requires inline Jinja2 (or offline encryption of keys) to generate the type 7 passwords required.  This PR automates this process.

Schema Changes:
```
underlay_ospf_authentication:
  enabled: true
  message_digest_keys:
    - id: 1
      hash_algorithm: sha512
      key: arista123
```

## How to test
Molecule tests added in `eos_designs_unit_tests` and `evpn_underlay_ospf_overlay_ebgp`.
Test manually by adding above YAML at fabric level with ospf configured as the underlay.

### Repository Checklist

- [X] My code has been rebased from devel before I start
- [X] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [X] My change requires a change to the documentation and documentation have been updated accordingly.
- [X] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+              - sha1
+              - sha256
+              - sha384
+              - sha512
+              default: sha512
+            key:
+              type: str
+              description: |
+                Key password.
+                Only plaintext passwords are supported here as the password will need to be encrypted for every underlay interface.
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml]: Are we sure we need this? The key could be set under the ospf process instead of per-interface. I think the process id is then used in the hash, so the user could leverage our encrypt filter to generate the correct hash.

Comment (jonxstill) [reply python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml]: As discussed just now, EOS only supports per-interface authentication, not per-process or per-area.

Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml]: Can we reword this to state that AVD is doing the job of encrypting the password? here it feels like `will need to be encrypted` implies extra work required.

Something like:
```suggestion
                Only plaintext passwords are supported here, the per-interface hashed password is automatically created by `eos_designs` using the relevant keys.
```

File: python-avd/pyavd/_eos_designs/structured_config/underlay/ethernet_interfaces.py [hunk 2]
```diff
@@ -99,6 +100,33 @@ def ethernet_interfaces(self: AvdStructuredConfigUnderlay) -> list | None:
...
+                            if not ("id" in ospf_key and "key" in ospf_key):
+                                continue
+
+                            ospf_keys.append(
+                                {
+                                    "id": ospf_key["id"],
+                                    "hash_algorithm": ospf_key.get("hash_algorithm", "sha512"),
+                                    "key": encrypt(
+                                        ospf_key["key"],
+                                        passwd_type="ospf_message_digest",
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/underlay/ethernet_interfaces.py]:
```suggestion
                                        passwd_type="ospf_message_digest",  # NOSONAR
```

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  underlay_ospf_authentication:
+    documentation_options:
+      table: ospf-settings
+    type: dict
+    keys:
+      enabled:
+        type: bool
+        default: false
+      message_digest_keys:
+        type: list
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml:19]: Should we add some min length and max length on this list? As far as I understand the doc, EOS supports one or two keys.
```suggestion
        type: list
        required: true
        primary_key: id
        min_length: 1
        max_length: 2
```

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+              type: str
+              valid_values:
+              - md5
+              - sha1
+              - sha256
+              - sha384
+              - sha512
+              default: sha512
+            key:
+              type: str
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml:41]:
```suggestion
              type: str
              required: true
              convert_types:
                - int
```

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+# Use Ctrl + Space to get suggestions for every field. Autocomplete will pop up after typing 2 letters.
+type: dict
+keys:
+  underlay_ospf_authentication:
+    documentation_options:
+      table: ospf-settings
+    type: dict
+    keys:
+      enabled:
+        type: bool
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml:15]:
```suggestion
        type: bool
        required: true
```

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/MLAG_OSPF_TESTS.yml [hunk 6]
```diff
@@ -1,5 +1,10 @@
 ---
 underlay_routing_protocol: ospf
+underlay_ospf_authentication:
+  enabled: true
+  message_digest_keys:
+    - id: 1
+      key: arista123arista123
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/MLAG_OSPF_TESTS.yml]: please add two keys here

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+              - sha256
+              - sha384
+              - sha512
+              default: sha512
+            key:
+              type: str
+              required: true
+              convert_types:
+              - int
+              description: |
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml]:
```suggestion
              description: |-
```

**Review by ClausHolbechArista — APPROVED** [review]
Looks good to me overall. Suggesting improving the test a bit for two keys but otherwise all good.

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+              valid_values:
+              - md5
+              - sha1
+              - sha256
+              - sha384
+              - sha512
+              default: sha512
+            key:
+              type: str
+              required: true
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/underlay_ospf_authentication_schema.yml:42]:
```suggestion
              required: true
              max_length: 16
              min_length: 1
```

**Review by gmuloc — APPROVED** [review]
LGTM
